src/interface/Variable.hpp

```diff
@@ -164,58 +106,70 @@ class Variable : public AthenaArray<T> {
-    if ( metadata.IsSet(Metadata::Sparse) ) {
-      throw std::invalid_argument ("Sparse not yet implemented for FaceVariable");
-    }
+  FaceVariable(const std::string label, const std::array<int,6> ncells,
+    const Metadata& metadata)
+    : data(label,ncells[5], ncells[4], ncells[3], ncells[2], ncells[1], ncells[0]),
+      _dims(ncells),
+      _m(metadata),
+      _label(label) {
+    assert ( !metadata.IsSet(Metadata::Sparse)
+             && "Sparse not implemented yet for FaceVariable" );
   }
 
   /// Create an alias for the variable by making a shallow slice with max dim
-  FaceVariable(std::string label, FaceVariable &src) :
-    FaceField(0,0,0,DATASTATUS::allocated),
-    _label(label),
-    _m(src.metadata()) {
-    int dim = 6;
-    int start = 0;
-    this->x1f.InitWithShallowSlice(src.x1f, dim, start, src.x1f.GetDim6());
-    this->x2f.InitWithShallowSlice(src.x2f, dim, start, src.x2f.GetDim6());
-    this->x3f.InitWithShallowSlice(src.x3f, dim, start, src.x3f.GetDim6());
-  }
+  FaceVariable(std::string label, FaceVariable<T> &src)
+    : data(src.data),
+      _dims(src._dims),
+      _m(src._m),
+      _label(label) { }
+
+  // KOKKOS_FUNCTION FaceVariable() = default;
+  // KOKKOS_FUNCTION FaceVariable(const FaceVariable<T>& v) = default;
+  // KOKKOS_FUNCTION ~FaceVariable() = default;
 
   ///< retrieve label for variable
-  std::string label() { return _label; }
+  const std::string& label() const { return _label; }
 
   ///< retrieve metadata for variable
-  Metadata metadata() { return _m; }
+  const Metadata metadata() const { return _m; }
 
   /// return information string
   std::string info();
 
   // TODO(JMM): should this be 0,1,2?
   // Should we return the reference? Or something else?
-  AthenaArray<Real>& Get(int i) {
-    if (i == 1) return (this->x1f);
-    if (i == 2) return (this->x2f);
-    if (i == 3) return (this->x3f);
-    throw std::invalid_argument("Face must be x1f, x2f, or x3f");
+  KOKKOS_FORCEINLINE_FUNCTION
+  ParArrayND<T>& Get(int i) {
+    assert( 1 <= i && i <= 3 );
+    if (i == 1)
+      return (data.x1f);
+    if (i == 2)
+      return (data.x2f);
+    else // i == 3
+      return (data.x3f);
   }
   template<typename...Args>
-  Real& operator()(int dir, Args... args) {
-    if (dir == 1) return x1f(std::forward<Args>(args)...);
-    if (dir == 2) return x2f(std::forward<Args>(args)...);
-    if (dir == 3) return x3f(std::forward<Args>(args)...);
-    throw std::invalid_argument("Face must be x1f, x2f, or x3f");
+  KOKKOS_FORCEINLINE_FUNCTION
+  T& operator()(int dir, Args... args) const {
+    assert( 1 <= dir && dir <= 3 );
+    if (dir == 1)
+      return data.x1f(std::forward<Args>(args)...);
+    if (dir == 2)
+      return data.x2f(std::forward<Args>(args)...);
+    else // dir == 3
+      return data.x3f(std::forward<Args>(args)...);
   }
 
+  bool isSet(const MetadataFlag bit) const { return _m.IsSet(bit); }
+
+  FaceArray<T> data;
+
  private:
+  std::array<int,6> _dims;
```

Comment:
I'm torn about storing `_dims`, as this information is contained within `ParArrayND data`. But there is arithmetic involved in extracting it, where here there's none.

Reply:
I removed it, but I'm second guessing myself.